.github/workflows/ci.yml

```diff
@@ -94,3 +94,39 @@ jobs:
         directory: ./build
         files: ./build/coverage.out
         fail_ci_if_error: true
+
+  format:
+    name: Formatting on ${{ matrix.platform.name }}
+    runs-on: ${{ matrix.platform.os }}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        platform:
+        - { name: Linux, os: ubuntu-latest }
+        - { name: macOS, os: macos-12 }
+
+    steps:
+    - name: Checkout Code
+      uses: actions/checkout@v2
+
+    - name: Install Linux Dependencies
+      if: runner.os == 'Linux'
+      run: sudo apt-get update && sudo apt-get install clang-format-12
+
+    - name: Install macOS Dependencies
+      if: runner.os == 'macOS'
+      run: brew install clang-format
+
+    - name: Format Code on Linux
+      if: runner.os == 'Linux'
+      run: cmake -DCLANG_FORMAT_EXECUTABLE=clang-format-12 -P $GITHUB_WORKSPACE/cmake/Format.cmake
+
+    - name: Format Code on macOS
+      if: runner.os == 'macOS'
+      run: cmake -DCLANG_FORMAT_EXECUTABLE=/usr/local/Cellar/clang-format/14.0.5/bin/clang-format -P $GITHUB_WORKSPACE/cmake/Format.cmake
+
+    - name: Check Formatting
+      run: |
+        git diff
+        exit `git status --porcelain | wc -l`
```

Comment:
Does this lead to a failure or just output somewhere the line count?

Reply:
It's saying "if changes are observed to the repo, return a non-zero exit code". It's some shell trickery I came up with a year or so ago. We need some way of returning non-zero if we detect that running the auto formatter produces changes to the codebase.

Reply:
There's also the [`--exit-code` argument](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---exit-code), which returns 0 if no differences, and 1 if differences are found.
You could use it as follows:
```sh
git diff --exit-code || exit 1
```
(or possibly add some error output). The problem with ```exit `cmd` ```  is that it exits unconditionally, so when people add another step in the future, they might be confused as to why it's not executed. `cmd || exit 1` avoids this.

**Note:** when using those commands in the past, I encountered quite a few subtleties regarding git-ignored and indexed/staged files; e.g. there's also the `--cached` argument.

Your approach with using the line count as exit code is pretty cool, I have to say 🙂

Reply:
> There's also the [--exit-code argument](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---exit-code), which returns 0 if no differences, and 1 if differences are found.

That's perfect! I didn't know that existed. I'll have to try that out later. Probably a good fit for this PR.

Reply:
Wow this is incredible. 
```
git diff
exit `git status --porcelain | wc -l`
```
becomes simply
```
git diff --exit-code
```
I love it! I used a separate branch to check that this fails when expected then pushed that change up to this branch.